introduce `implied_by` in `#[unstable]` attribute

davidtwco opened the pull request:
Requested by the library team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/better.20support.20for.20partial.20stabilizations/near/285581519).

If part of a feature is stabilized and a new feature is added for the remaining parts, then the `implied_by` meta-item can be added to `#[unstable]` to indicate which now-stable feature was used previously. 

```diagnostic
error: the feature `foo` has been partially stabilized since 1.62.0 and is succeeded by the feature `foobar`
  --> $DIR/stability-attribute-implies-using-unstable.rs:3:12
   |
LL | #![feature(foo)]
   |            ^^^
   |
note: the lint level is defined here
  --> $DIR/stability-attribute-implies-using-stable.rs:2:9
   |
LL | #![deny(stable_features)]
   |         ^^^^^^^^^^^^^^^
help: if you are using features which are still unstable, change to using `foobar`
   |
LL | #![feature(foobar)]
   |            ~~~~~~
help: if you are using features which are now stable, remove this line
   |
LL - #![feature(foo)]
   |
```

When a `#![feature(..)]` attribute still exists for the now-stable attribute, then there this has two effects:

- There will not be an stability error for uses of items from the implied feature which are still unstable (until the `#![feature(..)]` is removed or updated to the new feature). 
- There will be an improved diagnostic for the remaining use of the feature attribute for the now-stable feature.

```rust
        /// If part of a feature is stabilized and a new feature is added for the remaining parts,
        /// then the `implied_by` attribute is used to indicate which now-stable feature previously
        /// contained a item.
        ///
        /// ```pseudo-Rust
        /// #[unstable(feature = "foo", issue = "...")]
        /// fn foo() {}
        /// #[unstable(feature = "foo", issue = "...")]
        /// fn foobar() {}
        /// ```
        ///
        /// ...becomes...
        ///
        /// ```pseudo-Rust
        /// #[stable(feature = "foo", since = "1.XX.X")]
        /// fn foo() {}
        /// #[unstable(feature = "foobar", issue = "...", implied_by = "foo")]
        /// fn foobar() {}
        /// ```
```

In the Zulip discussion, this was envisioned as `implies` on `#[stable]` but I went with `implied_by` on `#[unstable]` because it means that only the unstable attribute needs to be changed in future, not the new stable attribute, which seems less error-prone. It also isn't particularly feasible for me to detect whether items from the implied feature are used and then only suggest updating _or_ removing the `#![feature(..)]` as appropriate, so I always do both.

There's some new information in the cross-crate metadata as a result of this change, that's a little unfortunate, but without requiring that the `#[unstable]` and `#[stable]` attributes both contain the implication information, it's necessary: 

```rust
    /// This mapping is necessary unless both the `#[stable]` and `#[unstable]` attributes should
    /// specify their implications (both `implies` and `implied_by`). If only one of the two
    /// attributes do (as in the current implementation, `implied_by` in `#[unstable]`), then this
    /// mapping is necessary for diagnostics. When a "unnecessary feature attribute" error is
    /// reported, only the `#[stable]` attribute information is available, so the map is necessary
    /// to know that the feature implies another feature. If it were reversed, and the `#[stable]`
    /// attribute had an `implies` meta item, then a map would be necessary when avoiding a "use of
    /// unstable feature" error for a feature that was implied.
```

I also change some comments to documentation comments in the compiler, add a helper for going from a `Span` to a `Span` for the entire line, and fix a incorrect part of the pre-existing stability attribute diagnostics.

cc @yaahc

davidtwco committed:
attr/passes: comment -> doc comment

Change some regular comments into documentation comments.

Signed-off-by: David Wood <david.wood@huawei.com>

davidtwco committed:
attr: fix expected meta-item for `#[stable]`

When an unexpected meta item is provided to `#[stable]`, the diagnostic
lists "since" and "note" as expected meta-items, however the surrounding
code actually expects "feature" and "since".

Signed-off-by: David Wood <david.wood@huawei.com>

davidtwco committed:
middle: add `implies_by` to `#[unstable]`

If part of a feature is stabilized and a new feature is added for the
remaining parts, then the `implied_by` attribute can be used to indicate
which now-stable feature previously contained a item. If the now-stable
feature is still active (if the user has only just updated rustc, for
example) then there will not be an stability error for uses of the item
from the implied feature.

Signed-off-by: David Wood <david.wood@huawei.com>

davidtwco committed:
span: add `span_extend_to_line` helper

Adds a simple helper function to the `SourceMap` for extending a `Span`
to encompass the entire line it is on - useful for suggestions where
removing a line is the suggested action.

Signed-off-by: David Wood <david.wood@huawei.com>

davidtwco committed:
passes: improved partial stabilization diagnostic

Improves the diagnostic when a feature attribute is specified
unnecessarily but the feature implies another (i.e. it was partially
stabilized) to refer to the implied feature.

Signed-off-by: David Wood <david.wood@huawei.com>

rust-highfive commented:
r? @michaelwoerister

(rust-highfive has picked a reviewer for you, use r? to override)

davidtwco commented on compiler/rustc_passes/src/stability.rs:
```diff
@@ -971,12 +985,41 @@ pub fn check_unused_or_stable_features(tcx: TyCtxt<'_>) {
+) {
+    tcx.struct_span_lint_hir(lint::builtin::STABLE_FEATURES, hir::CRATE_HIR_ID, span, |lint| {
+        lint.build(&format!(
+            "the feature `{feature}` has been partially stabilized since {since} and is succeeded \
```
I would have made these translatable given that's what I'm doing everywhere else but I have a patch coming for `rustc_passes` in the next few days so it can wait for that. :)

CAD97 commented:
What's the behavior of `implied_by = "still_unstable"` and `implied_by = "not_a_feature"`?

The former is useful for feature subsetting while still unstable, e.g. `min_specialization`. The latter could be useful for feature renaming, but should probably be prevented as likely an unintended typo.

davidtwco commented:
> What's the behavior of `implied_by = "still_unstable"` and `implied_by = "not_a_feature"`?
> 
> The former is useful for feature subsetting while still unstable, e.g. `min_specialization`. The latter could be useful for feature renaming, but should probably be prevented as likely an unintended typo.

`implied_by = "still_unstable"` works right now, and the behaviour is the exact same as `implied_by = "stable"`. This could be changed but it sounds like it might be useful to t-libs, so I don't mind leaving it as is.

As for `implied_by = "not_a_feature"`, I don't really know what would happen if you did this, I suspect nothing would happen (but we might crash, I haven't checked). If it is feasible, I'll add a check somewhere to prevent this.

bors commented:
:umbrella: The latest upstream changes (presumably #95956) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

michaelwoerister commented on compiler/rustc_middle/src/middle/stability.rs:
```diff
@@ -62,6 +62,19 @@ pub struct Index {
+    /// reported, only the `#[stable]` attribute information is available, so the map is necessary
+    /// to know that the feature implies another feature. If it were reversed, and the `#[stable]`
+    /// attribute had an `implies` meta item, then a map would be necessary when avoiding a "use of
+    /// unstable feature" error for a feature that was implied.
```
Very helpful docs 😍

michaelwoerister commented on compiler/rustc_metadata/src/rmeta/encoder.rs:
```diff
@@ -1767,6 +1775,13 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
+    fn encode_stability_implications(&mut self) -> LazyArray<(Symbol, Symbol)> {
+        empty_proc_macro!(self);
+        let tcx = self.tcx;
+        let implications = tcx.stability().implications.clone();
```
Also, why not use the `stability_implications` query?

davidtwco replied:
Just an oversight from an earlier iteration of the patch, fixed.

michaelwoerister commented on src/test/ui/stability-attribute/stability-attribute-implies-no-feature.rs:
```diff
@@ -0,0 +1,13 @@
+// aux-build:stability-attribute-implies.rs
+
+// Tests that despite the `foobar` feature being implied by now-stable feature `foo`, if `foo`
+// isn't allowed in this crate then an error will be emitted.
```
Should that be 

> ... by now-stable feature `foo`, if *`foobar`* isn't allowed ...

davidtwco replied:
Fixed!

michaelwoerister commented:
The implementation looks good to me. Thanks, @davidtwco!

Maybe someone from @rust-lang/libs wants to take another look to make sure that this PR actually implements the desired behavior? Otherwise, feel free to r=me.

michaelwoerister commented:
> As for implied_by = "not_a_feature", I don't really know what would happen if you did this, I suspect nothing would happen (but we might crash, I haven't checked). If it is feasible, I'll add a check somewhere to prevent this.

Adding a check like this sounds like a good idea. Here maybe?: https://github.com/rust-lang/rust/pull/99212/files#diff-5f57ad10e1bdde3d046b258d390fd2ecc6f1511158aa130cebb72093da16ef29R955-R958

m-ou-se commented:
cc @rust-lang/libs-api

davidtwco committed:
passes: check implied feature exists

Add a check confirming that features referenced in `implied_by` meta
items actually exist.

Signed-off-by: David Wood <david.wood@huawei.com>

michaelwoerister commented:
OK, r=me with the test fixed.

davidtwco commented:
@bors r=michaelwoerister

bors commented:
:pushpin: Commit e5872990d13abb088397e23b226439b1b4926b91 has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister e5872990d13abb088397e23b226439b1b4926b91 -->
<!-- homu: {"type":"Approved","sha":"e5872990d13abb088397e23b226439b1b4926b91","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

pnkfelix commented:
Hey @davidtwco , is there any chance that this PR was the cause of the performance regression reported in rollup PR #99520? You can see my line of reasoning about why I thought it might be, [over in zulip](https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/triage.202022.2007.2027/near/291056020).

davidtwco commented:
> Hey @davidtwco , is there any chance that this PR was the cause of the performance regression reported in rollup PR #99520? You can see my line of reasoning about why I thought it might be, [over in zulip](https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/triage.202022.2007.2027/near/291056020).

It's possible because of some additions to the cross-crate metadata. I've had another look at the changes and can't think of any ways to do less work that it currently does (it should just be writing a `HashMap` if there are elements in it, which there might only be one or two of, depends how much the library team are already using this).

yaahc commented:
I don't think we've landed any PRs yet that need this, only https://github.com/rust-lang/rust/pull/99573 which is still open and has another path to landing that we can use instead, so if there are perf issues with the impl as is and y'all feel like pumping the breaks that should be okay on our side for now at least.

steffahn commented:
Hi, out of curiosity, since I just found out this had actually been implemented, I wanted to try this on the only occurrence currently in std… and it doesn’t work at all?
```rs
#![feature(const_btree_new)]

use std::collections::BTreeMap;

const fn f() -> BTreeMap<(), ()> {
    let x = BTreeMap::new();
    let l = x.len();
    x
}
```
```
error: `BTreeMap::<K, V, A>::len` is not yet stable as a const fn
 --> src/lib.rs:7:13
  |
7 |     let l = x.len();
  |             ^^^^^^^
  |
  = help: add `#![feature(const_btree_len)]` to the crate attributes to enable
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f02c69b050aed07fbe2d7c0092146dee))

I’m not doing anything wrong, am I? I guess I should open an issue?

steffahn commented:
Oh… maybe it’s a problem because it isn’t using `unstable`/`stable` but instead `rustc_const_unstable`/`rustc_const_stable`!?
